python/tchannel/tornado/peer.py

```diff
@@ -525,160 +525,122 @@ def send(self, arg1, arg2, arg3,
 
-        peer = self.peer
-        connection = yield peer.connect()
-
-        request = Request(
-            service=self.service,
-            argstreams=[InMemStream(endpoint), arg2, arg3],
-            id=connection.next_message_id(),
-            headers=headers,
-            endpoint=endpoint,
-            ttl=ttl,
-            tracing=Trace(
-                name=endpoint,
-                trace_id=trace_id,
-                parent_span_id=parent_span_id,
-                endpoint=Endpoint(self.peer.host,
-                                  self.peer.port,
-                                  self.service),
-                traceflags=traceflag,
-            )
-        )
-
         # only retry on non-stream request
-        if request.is_streaming_request or self._hostport:
+        if Request.is_streaming_request(arg2, arg3):
             attempt_times = 1
+            ttl = 0
 
-        if request.is_streaming_request:
-            request.ttl = 0
-
-        # event: send_request
-        self.peer.tchannel.event_emitter.fire(
-            EventType.before_send_request,
-            request,
-        )
+        if self._hostport:
+            attempt_times = 1
 
-        try:
-            response = yield self.send_with_retry(
-                request, peer, attempt_times, retry_delay
-            )
-        except ProtocolError as protocol_error:
-            # event: after_receive_protocol_error
-            self.peer.tchannel.event_emitter.fire(
-                EventType.after_receive_system_error,
-                request,
-                protocol_error,
-            )
-            raise
-        except TimeoutError as operational_error:
-            # event: on_operational_error
-            self.peer.tchannel.event_emitter.fire(
-                EventType.on_operational_error,
-                request,
-                operational_error,
-            )
-            raise
-
-        log.debug("Got response %s", response)
-        # event: after_receive_response
-        self.peer.tchannel.event_emitter.fire(
-            EventType.after_receive_response,
-            request,
-            response,
+        response = yield self.send_with_retry(
+            endpoint, arg2, arg3, headers, trace_id,
+            parent_span_id, traceflag, ttl, attempt_times, retry_delay,
         )
 
         raise gen.Return(response)
 
     @gen.coroutine
-    def _send(self, connection, req):
-        response_future = connection.send_request(req)
-        with timeout(response_future, req.ttl):
-            response = yield response_future
-
-        raise gen.Return(response)
-
-    @gen.coroutine
-    def send_with_retry(self, request, peer, attempt_times, retry_delay):
+    def send_with_retry(self, endpoint, arg2, arg3, headers,
+                        trace_id, parent_span_id, traceflag,
+                        ttl, attempt_times, retry_delay):
         # black list to record all used peers, so they aren't chosen again.
```

Comment:
Can you add an assert here?

    assert attempt_times > 0

Because if attempt_times is 0, we will silently not make a call, and that's programmer error. We need that to fail.